### PR TITLE
[Feature] Prepare Lance scan optimizer operators and column pruning

### LINKED ISSUE #74665: [Feature] Native Lance connector for StarRocks
- [x] https://github.com/StarRocks/starrocks/pull/74659
## Is your feature request related to a problem? Please describe.

StarRocks currently supports external table connectors for Iceberg, Delta Lake, Hudi, Paimon, and other formats. However, there is no support for [Lance](https://github.com/lancedb/lance), a modern columnar data format optimized for ML workloads — embedding vectors, multi-modal data, and large-scale training datasets.

Users who store datasets in Lance format must currently ETL data into a supported format before querying with StarRocks, adding latency and operational complexity.

## Describe the solution you'd like

A native Lance connector that allows StarRocks to directly query Lance datasets on local and cloud filesystems (S3, ADLS Gen2, GCS), enabling:

- Ad-hoc exploration of ML training datasets (embeddings, document corpora)
- Quality analysis by joining Lance datasets with metadata in Iceberg/Hive tables
- Interactive analytics replacing Spark read paths on Lance tables

## Implementation Plan

The implementation is split into independently reviewable PRs:

| PR | Scope | Status |
|----|-------|--------|
| #74659 | Thrift definitions for Lance scan range fields | Open |
| #74660 | FE optimizer operators and rules | Open |
| #74661 | FE scan node and plan fragment builder | Open |
| #74662 | Java-extensions Lance JNI reader | Open |
| #74663 | BE optional pluggable connector module (follows benchmark pattern) | Open |
| #74664 | Build flag `--without-connector-lance` | Open |

**Already merged (Phase 1):** FE catalog & metadata discovery (`LanceConnector`, `LanceMetadata`, `LanceApiConverter`, `LanceTable`)

## Architecture

- **FE**: `LANCE` registered as `ConnectorType` and `TableType`. Query planning reuses `THdfsScanRange` with Lance-specific fields. CBO statistics integration.
- **BE**: Optional pluggable module following the benchmark connector pattern (`--without-connector-lance` compile-time opt-out). JNI scanner delegates to `lance-core` Java SDK.
- **java-extensions**: `LanceSplitScanner` using Arrow vectors via `lance-core:7.0.0`

## Future Work

- Predicate pushdown expression visitor
- Cloud storage credential passthrough (Azure Workload Identity, S3 IAM, GCS)
- HMS-backed catalog discovery (replacing property-based bootstrap)
- Native Rust FFI scanner (zero-copy via Arrow C Data Interface)

## Are you willing to submit PR?

Yes — PRs are already submitted (see table above).

### PR DESCRIPTION
**Pull request by beinan**
## Why I'm doing:

Lance scans need optimizer operators and rewrite rules before their SQL execution path can be enabled.

## What I'm doing:

Add logical/physical scan operators, implementation and statistics hooks, column pruning, and LIMIT merging. Clone predicate state when building or converting scans so later rewrites cannot mutate an earlier operator.

This is a preparatory layer. `LanceTable.isSupported()` remains false, and the SQL relation transformer is wired only in #74663, together with executable scans. An unexpected physical translation fails explicitly instead of recursing through the fallback visitor. The executable fragment translator follows in #74661.

Series order: #74660 → #74661 → #74662 → #74663 → #74664. Catalog and Thrift prerequisites (#74420 and #74659) are already merged. These branches are cumulative against `main`; each preceding layer is required.

### Validation

Focused optimizer and metadata tests pass at this PR head. The completed series also passes projection/filter/LIMIT and two-table join/aggregation planning tests.

## What type of PR is this:

- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
  - [ ] This pr needs auto generate documentation
- [ ] This is a backport pr

Documentation for user-visible execution is included in #74663; this series targets `main` only.

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 4.1
  - [ ] 4.0
  - [ ] 3.5

## Related Issue

Fixes #74665

This PR implements one layer of the linked feature. The complete implementation requires all five PRs in the series listed above.

### CODE REVIEW

**Comment by CelerData-Reviewer** [comment]
@codex review

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/transformer/RelationTransformer.java [hunk 1]
```diff
@@ -734,6 +735,9 @@ public LogicalPlan visitTable(TableRelation node, ExpressionMapping context) {
         } else if (Table.TableType.KUDU.equals(node.getTable().getType())) {
             scanOperator = new LogicalKuduScanOperator(node.getTable(), colRefToColumnMetaMapBuilder.build(),
                 columnMetaToColRefMap, Operator.DEFAULT_LIMIT, null);
+        } else if (Table.TableType.LANCE.equals(node.getTable().getType())) {
+            scanOperator = new LogicalLanceScanOperator(node.getTable(), colRefToColumnMetaMapBuilder.build(),
+                columnMetaToColRefMap, Operator.DEFAULT_LIMIT, null);
```
[the remark this answers is not in the document]
Comment (beinan) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/transformer/RelationTransformer.java:738]: False positive: `LanceTable.isSupported()` is changed to `true` in PR #74661 which is part of this series. See #74665 for the full dependency chain.

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/physical/PhysicalLanceScanOperator.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    }
+
+    @Override
+    public <R, C> R accept(OperatorVisitor<R, C> visitor, C context) {
+        return visitor.visitPhysicalLanceScan(this, context);
+    }
+
+    @Override
+    public <R, C> R accept(OptExpressionVisitor<R, C> visitor, OptExpression optExpression, C context) {
+        return visitor.visitPhysicalLanceScan(optExpression, context);
```
[the remark this answers is not in the document]
Comment (beinan) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/physical/PhysicalLanceScanOperator.java:50]: False positive: `PlanFragmentBuilder.visitPhysicalLanceScan` is added in PR #74661 which depends on this PR. See #74665 for the full dependency chain.

**Comment by beinan** [comment]
@github-actions run-checks

**Comment by beinan** [comment]
@github-actions run-checks

**Comment by beinan** [comment]
@github-actions run-checks

**Comment by beinan** [comment]
Hi @stephen-shelby @HangyuanLiu, gentle ping on this PR when you have a chance. It is the first (bottom) PR of the Lance connector stack (#74660 → #74661 → #74662 → #74663 → #74664) and only adds the optimizer operators/rules. CI is green except for a known flaky `PublishVersionDaemonTest.testSharedNothingPublishWithThreadPool` (unrelated to this change), which I have just re-triggered. Would appreciate your review to get the stack moving. Thanks!

**Comment by beinan** [comment]
@stephen-shelby @HangyuanLiu Conflicts have been resolved and this PR has been rebased onto the latest main. The stack is #74660 -> #74661 -> #74662 -> #74663 -> #74664. Could you please take another look when available? CI is rerunning on the updated commits.

**Comment by beinan** [comment]
@stephen-shelby @HangyuanLiu The dependency-chain review comments have been clarified and all review threads are now resolved. This PR is currently mergeable. Please take another look when available.

**Comment by beinan** [comment]
Rebased onto the latest main (conflict-free). CI was already green on this PR; re-running on the updated commits to keep the stack consistent.

**Comment by CelerData-Reviewer** [comment]
@codex review

**Comment by beinan** [comment]
@stephen-shelby @HangyuanLiu — this is the **bottom of the Lance connector stack** and nothing else can land until it does, so it is the one that needs your eyes first.

Status: **CI is fully green**, no conflicts, rebased onto latest main. The two Codex P1 comments were dependency-chain false positives (the referenced symbols land in #74661) and are resolved.

Scope is deliberately small — it only adds the optimizer operators and rules (`LogicalLanceScanOperator`, `PhysicalLanceScanOperator`, `LanceScanImplementationRule`) plus their unit tests. No behavior change to existing paths.

Merge order for the stack: **#74660 → #74661 → #74662 → #74663 → #74664**.

Would appreciate a review when you have a moment.

**Comment by alvin-phoenix-ai** [comment]
@mergify rebase

**Comment by alvin-phoenix-ai** [comment]
@mergify update

**Comment by beinan** [comment]
Hi @stephen-shelby @HangyuanLiu, a gentle ping when you have a chance to review this PR. CI is now green, and all previous review threads have been addressed and resolved. This is the first pending PR in the Lance connector series; it adds the optimizer groundwork while keeping Lance SQL queries disabled until the execution layers land. Your review would help move the series forward. Thank you!

**Comment by beinan** [comment]
Re-run the workflow
